frontends/p4/typeChecking/typeChecker.h

```diff
@@ -83,8 +83,21 @@ class TypeInference : public Transform {
     void setType(const IR::Node* element, const IR::Type* type);
     void setLeftValue(const IR::Expression* expression)
     { typeMap->setLeftValue(expression); }
+    bool isArrayIndexRuntime(const IR::Expression* e) const {
+        if (auto mem = e->to<IR::Member>()) {
+            if (auto ai = mem->expr->to<IR::ArrayIndex>()) {
+                auto right = ai->right;
+                if (!right->is<IR::Constant>()) {
+                    return true;
+                }
+            }
+        }
+        return false;
+    }
+
     bool isLeftValue(const IR::Expression* expression) const
-    { return typeMap->isLeftValue(expression) || expression->is<IR::DefaultExpression>(); }
+    { return typeMap->isLeftValue(expression) || expression->is<IR::DefaultExpression>()
```

Comment:
This is incorrect; the definition of left values is very clear in the spec.

Reply:
I searched in the p4-16 spec and do not see LeftValue in regards to AssignmentStatement.

Could you please do me a favor and use this URL below to show me where in the spec is the definition of values.

https://github.com/p4lang/p4-spec/blob/master/p4-16/spec/P4-16-spec.mdk

Reply:
This might be the section Mihai is thinking of: https://p4.org/p4-spec/docs/P4-16-v1.2.0.html#sec-lvalues

That describes the syntax and the meaning of expressions that can appear as an L-value.

How that ought to be implemented in the compiler to enable an expression like `hdr.vector[hdr.vector[hdr.ethernet.etherType[2:0]].e].e` to be allowed on the left hand side of an assignment, I have no idea.

Reply:
Thanks @jafingerhut 

Maybe the code is correct.

This is text from the same L-Value section of the spec.

```
The following is a legal l-value: `headers.stack[4].field`
```
A constant array index is allowed in the above expression.  Ours is a runtime index expression.  Why should ours not be allowed?  The only reason I can think of is that expression count is greater than zero.

Additionally, I was also looking at the grammar in p4c/frontend/p4/parsers/p4/p4parser.ypp and clearly see this grammar:

```
lvalue
    : prefixedNonTypeName                { $$ = new IR::PathExpression($1); }
    | THIS                               { $$ = new IR::This(@1); }  // experimental
    | lvalue "." name                    { $$ = new IR::Member(@1 + @3, $1, *$3); }
    | lvalue "[" expression "]"          { $$ = new IR::ArrayIndex(@1 + @4, $1, $3); }
    | lvalue "[" expression ":" expression "]" { $$ = new IR::Slice(@1 + @6, $1, $3, $5); }
    ;
```

A lvalue can include an ArrayIndex.